### PR TITLE
Fix focus for user actionmenu

### LINKED ISSUE #19464: Not possible to disable, delete etc a user via a screen reader
### Steps to reproduce
1. Navigate to users list.
2. Use arrow keys (or Control Option Right/ Control Option Left if on Mac OS) to move screen reader cursor to name of user.

### Expected behaviour
Popup menu should appear when name of user is interacted with via keyboard. Screen reader should indicate that the item will produce a popover menu.

### Actual behaviour
Nothing happens. The menu is triggered only by a mouseover event, which cannot be replicated by a screen reader.

## Server configuration detail

**Operating system:** Linux 3.10.0-1062.12.1.el7.x86_64 #1 SMP Tue Feb 4 23:02:59 UTC 2020 x86_64

**Webserver:** nginx/1.16.1 (fpm-fcgi)

**Database:** Mariadb 10.4.12

**PHP version:** 

7.3.14
Modules loaded: Core, date, libxml, openssl, pcre, zlib, filter, hash, Reflection, SPL, session, standard, cgi-fcgi, bz2, calendar, ctype, curl, dom, mbstring, fileinfo, ftp, gd, gettext, iconv, imap, intl, json, exif, mysqlnd, PDO, Phar, posix, shmop, SimpleXML, sockets, sqlite3, sysvmsg, sysvsem, sysvshm, tokenizer, xml, xmlwriter, xsl, mcrypt, mysqli, pdo_mysql, pdo_sqlite, wddx, xmlreader, igbinary, imagick, memcache, msgpack, zip, memcached, redis, Zend OPcache

**Nextcloud version:** 17.0.3 - 17.0.3.1

**Updated from an older Nextcloud/ownCloud or fresh install:** Updated from version as of early October 2019 (do not remember exact version number)

**Where did you install Nextcloud from:** Direct download from nextcloud.com

<details><summary>Signing status</summary>

Array
(
)

</details>

<details><summary>List of activated apps</summary>

```
Enabled:
 - accessibility: 1.3.0
 - activity: 2.10.1
 - audioplayer: 2.9.0
 - calendar: 2.0.1
 - cloud_federation_api: 1.0.0
 - comments: 1.7.0
 - contacts: 3.1.8
 - dav: 1.13.0
 - federatedfilesharing: 1.7.0
 - federation: 1.7.0
 - files: 1.12.0
 - files_pdfviewer: 1.6.0
 - files_rightclick: 0.15.1
 - files_sharing: 1.9.0
 - files_trashbin: 1.7.0
 - files_versions: 1.10.0
 - files_videoplayer: 1.6.0
 - firstrunwizard: 2.6.0
 - gallery: 18.4.0
 - issuetemplate: 0.6.0
 - logreader: 2.2.0
 - lookup_server_connector: 1.5.0
 - mail: 1.1.2
 - nextcloud_announcements: 1.6.0
 - notifications: 2.5.0
 - oauth2: 1.5.0
 - password_policy: 1.7.0
 - phonetrack: 0.6.2
 - privacy: 1.1.0
 - provisioning_api: 1.7.0
 - ransomware_detection: 0.6.0
 - ransomware_protection: 1.5.1
 - recommendations: 0.5.0
 - serverinfo: 1.7.0
 - spreed: 7.0.2
 - support: 1.0.1
 - survey_client: 1.5.0
 - systemtags: 1.7.0
 - text: 1.1.1
 - theming: 1.8.0
 - twofactor_backupcodes: 1.6.0
 - updatenotification: 1.7.0
 - viewer: 1.2.0
 - workflowengine: 1.7.0
Disabled:
 - admin_audit
 - encryption
 - files_external
 - sharebymail
 - user_ldap

```
</details>

<details><summary>Configuration (config/config.php)</summary>

```
{
    "passwordsalt": "***REMOVED SENSITIVE VALUE***",
    "secret": "***REMOVED SENSITIVE VALUE***",
    "trusted_domains": [
        "cloud.caturria.ca"
    ],
    "datadirectory": "***REMOVED SENSITIVE VALUE***",
    "dbtype": "mysql",
    "version": "17.0.3.1",
    "overwrite.cli.url": "https:\/\/cloud.caturria.ca",
    "dbname": "***REMOVED SENSITIVE VALUE***",
    "dbhost": "***REMOVED SENSITIVE VALUE***",
    "dbport": "",
    "dbtableprefix": "oc_",
    "mysql.utf8mb4": true,
    "dbuser": "***REMOVED SENSITIVE VALUE***",
    "dbpassword": "***REMOVED SENSITIVE VALUE***",
    "token_auth_enforced": true,
    "memcache.local": "\\OC\\Memcache\\Redis",
    "redis": {
        "host": "***REMOVED SENSITIVE VALUE***",
        "port": 6379,
        "timeout": 0,
        "password": "***REMOVED SENSITIVE VALUE***",
        "dbindex": 0
    },
    "installed": true,
    "instanceid": "***REMOVED SENSITIVE VALUE***",
    "maintenance": false,
    "mail_smtpmode": "smtp",
    "mail_smtpsecure": "tls",
    "mail_sendmailmode": "smtp",
    "mail_from_address": "***REMOVED SENSITIVE VALUE***",
    "mail_domain": "***REMOVED SENSITIVE VALUE***",
    "mail_smtpauthtype": "PLAIN",
    "mail_smtpauth": 1,
    "mail_smtphost": "***REMOVED SENSITIVE VALUE***",
    "mail_smtpport": "587",
    "mail_smtpname": "***REMOVED SENSITIVE VALUE***",
    "mail_smtppassword": "***REMOVED SENSITIVE VALUE***",
    "theme": "",
    "loglevel": 2,
    "simpleSignUpLink.shown": false
}
```
</details>

**Are you using external storage, if yes which one:** 

**Are you using encryption:** 

**Are you using an external user-backend, if yes which one:** 

## Client configuration

**Browser:** Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_2) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/13.0.4 Safari/605.1.15

**Operating system:** Mac OS Catalina

## Logs

<details><summary>Web server error log</summary>

```
Insert your web server log here 
```
</details>

<details><summary>Nextcloud log</summary>

```
Insert your Nextcloud log here
```
</details>

<details><summary>Browser log</summary>

Insert your browser log here, this could for example include:

	a) The javascript console log
	b) The network log
	c) ...
</details>

### PR DESCRIPTION
**Pull request by GretaD**
fixes #19464

### CODE REVIEW

File: apps/settings/src/components/UserList/UserRowSimple.vue [hunk 1]
```diff
@@ -65,7 +65,10 @@
 					</ActionButton>
 				</Actions>
 				<div class="userPopoverMenuWrapper">
-					<div v-click-outside="hideMenu" class="icon-more" @click="$emit('toggleMenu')" />
+					<a href="#"
+						v-click-outside="hideMenu"
+						class="icon icon-more"
+						@click="$emit('toggleMenu')" />
```
Comment (ChristophWurst) [inline apps/settings/src/components/UserList/UserRowSimple.vue:71]: doesn't this navigate? Have you tried `@click.prevent`?

Comment (GretaD) [reply apps/settings/src/components/UserList/UserRowSimple.vue:71]: i did try and nothing changed. What is happening now is that the icon-more is disappeared

**Comment by jancborchardt** [comment]
@caturria is it possible for you to test this as it fixes the issue you reported at https://github.com/nextcloud/server/issues/19464 ? Thank you!

File: apps/settings/src/components/UserList/UserRowSimple.vue [hunk 2]
```diff
@@ -65,7 +65,11 @@
 					</ActionButton>
 				</Actions>
 				<div class="userPopoverMenuWrapper">
-					<div v-click-outside="hideMenu" class="icon-more" @click="$emit('toggleMenu')" />
+					<button
+						v-click-outside="hideMenu"
+						class="icon icon-more"
```
Comment (jancborchardt) [inline apps/settings/src/components/UserList/UserRowSimple.vue:70]: We only need the "icon-more" class, the single "icon" class is deprecated.

Comment (GretaD) [reply apps/settings/src/components/UserList/UserRowSimple.vue:70]: Yes, I forgot to remove it. But did you test it, is it ok how it looks now?

**Comment by caturria** [comment]
Hi,
@@GretaD I will spin up a test instance using your branch this afternoon and let you know by tomorrow.
Thank you very much for taking care of this.

**Comment by caturria** [comment]
Hi,
Sorry for the delay.
Apologies for asking, but if I could get a little guidance on upgrading a backup of my instance to this branch, it would be much appreciated.
I have followed all of the steps here:
https://docs.nextcloud.com/server/latest/admin_manual/maintenance/manual_upgrade.html
(Disabled as many apps as possible, backed up files and database, cloned branch and renamed to original Nextcloud directory, copied config/config.php from old to new, execute php occ upgrade).
It fails indicating that the Viewer app’s files have not been correctly replaced.
Viewer cannot be disabled, and apps/viewer does not exist in my old Nextcloud instance directory.
Kind regards,

Jordan.

**Comment by GretaD** [comment]
> (Disabled as many apps as possible, backed up files and database, cloned branch and renamed to original Nextcloud directory, copied config/config.php from old to new, execute php occ upgrade). It fails indicating that the Viewer app’s files have not been correctly replaced. Viewer cannot be disabled, and apps/viewer does not exist in my old Nextcloud instance directory. 

Hey Jordan, does the viewer app exist in your apps? If yes, maybe try to delete the app and clone it again: https://github.com/nextcloud/viewer.
It is ok to ask more questions if something else pops up. So dont worry :)

**Comment by caturria** [comment]
Hi,
Alright; I got the update installed.
What’s working better:

The users list itself reads well now. Scrolling through the options using Control Option Right, here’s what I hear:
User Name

Display Name
Email

Groups

Quota

<Name of first user>

<Display name of first user>

<Email of first user>

0B, 181.2GB used

90% progress indicator

Edit user button.
Where problems remain:
1. The administration manual indicates that a mouse hover over a user would reveal an actions menu which offers choices such as to delete or disable a user. There still does not appear to be a way to access this using the keyboard/ screen reader.

2. The ‘Edit User’ button sometimes does not have its name spoken (sometimes it simply reads as ‘Button’). If it reads as ‘Button’, moving the screen reader cursor away from it and then back will cause it to read as ‘Edit user button’ the second time. Does the caption ‘Edit User’ appear dynamically in response to a mouse over event or something to that effect? If so it would need Aria attributes set in order to consistently speak its purpose.
3. The user editing screen is problematic in the following ways:
*Still has focus issues, on Mac Voiceover still experiences the broken record effect where it gets stuck in a loop between a few options instead of tracking the form elements from beginning to end;
*Does not  focus the correct options when edit user is pressed.
If I press ’Add a new User’, it correctly focuses the user name field and the form is for the most part accessible, however I cannot set custom quotas (the ’Select user quota’ field does not consistently grab input focus, and if I do get it to accept input and type 50GB for instance, the text disappears as soon as it loses focus. So I only have Default quota, Unlimited, 1GB, 5GB and 10GB to choose from, with no way of knowing which of those has actually been chosen.
In summary:
*I can read the users list correctly,
*I cannot delete or disable a user,
*I cannot edit a user without significant difficulty,
*I can create a new user, but cannot assign their quotas.
Thank you very much for your attention to this matter and for your work towards resolving it. Please stay safe and healthy.

Kind regards,

Jordan.

**Comment by GretaD** [comment]
Thank you @caturria for the detailed explanation, thats very helpful. So i will move on and fix the action menu part, and we open another ticket for quota, ok?

**Comment by caturria** [comment]
Hi,
Are you asking me to open another issue for the quota, or saying you’ve done it at your end?

Kind regards,

Jordan.

**Comment by GretaD** [comment]
> Hi, Are you asking me to open another issue for the quota, or saying you’ve done it at your end? Kind regards, Jordan.

Yes please, you open it, i wanted to open it coping what you have wrote on the comment above, but didnt have the time, plus its better if you open it since you have all the details :)

File: apps/settings/src/components/UserList/UserRowSimple.vue [hunk 3]
```diff
@@ -65,7 +65,11 @@
...
+						:aria-label="t('settings', 'Action menu')"
```
Comment (skjnldsv) [inline apps/settings/src/components/UserList/UserRowSimple.vue:71]:
```suggestion
						:aria-label="t('settings', 'Toggle user actions menu')"
```

File: apps/settings/src/components/UserList/UserRowSimple.vue [hunk 4]
```diff
@@ -65,7 +65,11 @@
...
+						@click.prevent="$emit('toggleMenu')" />
 					<div class="popovermenu" :class="{ 'open': openedMenu }">
```
Comment (skjnldsv) [inline apps/settings/src/components/UserList/UserRowSimple.vue:73]:
```suggestion
					<div class="popovermenu" :class="{ 'open': openedMenu }" aria-expanded="openedMenu">
```

Comment (GretaD) [reply apps/settings/src/components/UserList/UserRowSimple.vue:73]: `aria-expanded` is Boolean, so its true or false, and in our case is true, right? And not `openedMenu`

Comment (skjnldsv) [reply apps/settings/src/components/UserList/UserRowSimple.vue]: Right, should be :aria-expanded="openedMenu"

Comment (GretaD) [reply apps/settings/src/components/UserList/UserRowSimple.vue]: I have never seen in this form, also phpstorm complains "wrong attribute value"

Comment (skjnldsv) [reply apps/settings/src/components/UserList/UserRowSimple.vue]: What do you mean You're never seen this form?
This is vue props binding?

Comment (GretaD) [reply apps/settings/src/components/UserList/UserRowSimple.vue]: i mean i havent like this:  :aria-expanded="openedMenu", i have only seen it like this:  :aria-expanded="true/false"

Comment (skjnldsv) [reply apps/settings/src/components/UserList/UserRowSimple.vue]: But openedMenu is a variable that is true or false :p
https://github.com/nextcloud/server/blob/3c76ebe13dacdd17585898e822077da8bc6c7e4c/apps/settings/src/components/UserList/UserRowSimple.vue#L119-L122

**Comment by GretaD** [comment]
ready for review and testing @skjnldsv @jancborchardt

**Comment by skjnldsv** [comment]
Cannot review yet because we need #20591

**Comment by GretaD** [comment]
> Cannot review yet because we need #20591

its merged, please check it out

**Comment by ChristophWurst** [comment]
> its merged, please check it out

You have to update this branch :)

**Comment by GretaD** [comment]
i need to fix the border and background color before final review.` Icon more` has inherited the button characteristics 
 
iconmore-buton

**Comment by GretaD** [comment]
This is a bit tricky, somehow the button is focused but I cannot change the hover opacity of the` icon-more`. It looks like was a z-index problem because when focus via tab, the color is pale, but didnt work.
I added button:focus temporarily to test the tab focus, but have no idea how to add the opacity for icon button.
ICON-MORE_OPEN

File: apps/settings/src/components/UserList/UserRowSimple.vue [hunk 5]
```diff
@@ -170,10 +174,19 @@ export default {
...
+}
+	.icon-more {
+		background-color: var(--color-main-background);
+		background-image: var(--icon-more-000);
+		border: 0;
+	}
+	//not final solution
+	.button:focus {
+		background-color: var(--color-background-dark)
 	}
```
Comment (skjnldsv) [inline apps/settings/src/components/UserList/UserRowSimple.vue:191]:
```suggestion
	.userPopoverMenuWrapper .button.icon-more {
		&:hover,
		&:focus,
		&:active {
			background-color: var(--color-background-dark)
		}
	}
	
```

Comment (skjnldsv) [reply apps/settings/src/components/UserList/UserRowSimple.vue:191]: Ah no, thi sis because the css is coming from /apps/settings/css/settings.scss:1679.

The apps/users page isn't using scoped styling in components. We weren't sure if it was the good call back then :)

Comment (GretaD) [reply apps/settings/src/components/UserList/UserRowSimple.vue:191]: It doesnt make any difference for what is missing
and what is missing is that when you tab, the 3 dot menu has no opacity, when you hover with mouse, it does(because of settings.scss:1679 as you said) :/

Comment (skjnldsv) [reply apps/settings/src/components/UserList/UserRowSimple.vue:191]: So change it in the settings.scss:1679 ? :stuck_out_tongue_winking_eye:

Comment (GretaD) [reply apps/settings/src/components/UserList/UserRowSimple.vue]: right :) ready for review then

File: apps/settings/src/components/UserList/UserRowSimple.vue [hunk 6]
```diff
@@ -170,10 +174,23 @@ export default {
...
+		background-image: var(--icon-more-000);
+		border: 0;
+	}
+	//not final solution
+	.userPopoverMenuWrapper .button.icon-more {
+		&:hover,
+		&:focus,
+		&:active {
+			background-color: var(--color-background-dark)
+		}
```
Comment (skjnldsv) [inline apps/settings/src/components/UserList/UserRowSimple.vue:194]:
```suggestion
```

**Review by skjnldsv — APPROVED** [review]

File: apps/settings/src/components/UserList/UserRowSimple.vue [hunk 7]
```diff
@@ -170,10 +174,15 @@ export default {
...
-<style scoped>
+<style lang="scss">
 	.cellText {
 		overflow: hidden;
 		text-overflow: ellipsis;
 		white-space: nowrap;
+}
+	.icon-more {
+		background-color: var(--color-main-background);
+		background-image: var(--icon-more-000);
```
Comment (skjnldsv) [inline apps/settings/src/components/UserList/UserRowSimple.vue:185]: Why do you need to specify the background?
Should be covered by the icon-more class already.

Seeing the changes in apps/settings/css/settings.scss, I don't see the need to have any rules in the vue style section here.
Border make sense though

**Review by skjnldsv — CHANGES_REQUESTED** [review]

File: apps/settings/src/components/UserList/UserRowSimple.vue (hunk 7 above)
Comment (GretaD) [reply apps/settings/src/components/UserList/UserRowSimple.vue:185]: background-color should be there because it inherits the button ` background-color: var(--color-background-dark);` and we dont want that.

the image is taken from settings, you're right.

**Review by skjnldsv — APPROVED** [review]

**Review by ChristophWurst — APPROVED** [review]

**Comment by ChristophWurst** [comment]
ah this needs squashing and a rebase

**Review by jancborchardt — APPROVED** [review]
Tested and works!
